mitxgraders/listgrader.py

```diff
@@ -758,9 +756,18 @@ def check_response(self, answer, student_input, **kwargs):
         else:
             input_list = find_optimal_order(checker, pad_ans, pad_stud)
 
-        result = consolidate_cfn_return(input_list,
-                                        n_expect=len(answers),
-                                        partial_credit=self.config['partial_credit'])
+        # Consolidate the separate results into a single result
+        result = consolidate_single_return(input_list,
+                                           n_expect=len(answers),
+                                           partial_credit=self.config['partial_credit'])
+
+        # If the result is ok, display the message for this answer
+        if result['ok'] is True and msg != '':
```

Comment:
This is the only thing I'm in two minds about. I'd like the message to display if the student responses matched their expect values, and I'm using 'ok' as a proxy for this. However, if grade_decimal for a response is less than 1, then I won't catch it. There's just no way of reporting that the entire answer matched something, irregardless of the grade_decimal that was assigned to it... Any thoughts?

Reply:
Something similar to this. You want `nice job!` to display for `cat, dog` and `cat, DOG`, even though `dog` doesn't quite get full credit. You can't associate a message with `dog` or `DOG` directly, because then `unicorn, dog` and `unicorn, DOG` will both get that message.

Reply:
In the  situation above, if I wanted to give a message on “cat, DOG” but not “unicorn, DOG” then I  would use a tuple of answers in the outermost SingleListGrader. 

Does that make sense? Sorry I can’t give a code example now, on my phone

Reply:
Here's what I have in mind exactly:
```python
grader = SingleListGrader(
    answers=(
        {
            'expect': [('X', {'expect': 'x', grade_decimal:0.5}),
                       ('Y', {'expect': 'y', grade_decimal:0.5})], 
            'msg': 'Good pairing!'
        },
        {
            'expect': [('W', {'expect': 'w', grade_decimal:0.5}),
                       ('Z', {'expect': 'z', grade_decimal:0.5})], 
            'msg': 'Good pairing!'
        }
    ),
    subgrader=StringGrader()
)
```
The message should appear for `X, Y`, `x, Y`, `X, y`, and `x, y`, or similarly for `W/w` and `Z/z`. If you get the pairing wrong though, say, `X, Z`, you don't get a message.

Currently, the message will only appear for `X, Y` or `W,Z`, because those are the only combination that awards full credit.

Happy to wait for you to take a look tonight - go and enjoy your day!

Reply:
Ah, I understand. Well, to your original question:

> This is the only thing I'm in two minds about. I'd like the message to display if the student responses matched their expect values, and I'm using 'ok' as a proxy for this. However, if grade_decimal for a response is less than 1, then I won't catch it. There's just no way of reporting that the entire answer matched something, irregardless of the grade_decimal that was assigned to it... Any thoughts?

I agree that would be a more logical way to show messages.  I think that `input_list` (before consolidation) contains all the information you need. Something like

```python
if all(r['grade_decimal']>0 for r in input_list) and msg != '':
    result['msg'] = msg if result['msg'] == '' else result['msg'] + '\n' + msg
```
(or check that r['ok'] is truthy should be fine, too)

Reply:
I considered this, but I think it fails for nested `SingleListGrader`s, where partial credit might just mean that you got one wrong?